libraries/libfc/include/fc/io/random_access_file.hpp

```diff
@@ -0,0 +1,465 @@
+ threads, but an individual read/write_datastream must only be used by a single thread at a time. unpack_from() and pack_to() are just
+ simple helper functions based on these read/write_datastreams. Be aware that read/write_datastreams are buffered. Write buffers are flushed
+ as needed, both upon destruction of the datastream and also when an internal buffer becomes full.
+
+seekable_device() may be called from multiple threads simultaneously. This returns a Boost Iostreams SeekableDevice. Similar to datastreams,
+ multiple devices created from the same random_access_file can be used simultaneously from different threads, but an individual
+ device must only be used by a single thread at a time. These read/write_devices are _not_ buffered.
+
+There is a another important interaction between random_access_file and the read/write_datastream/device that a file creates: it is safe to use
+ the datastream/device after the random_access_file they were created from has been destroyed. For example, the following is considered safe:
+   std::optional<random_access_file> thefile = random_access_file("somefile");
+   write_datstream ds = thefile.write_ds(42);
+   thefile.reset();
+   fc::raw::pack(ds, (uint32_t)0xbeef);
+
+size(), resize(), and punch_hole() may be called from multiple threads simultaneously. Other threads performing reads or writes on affected
+ ranges will give undefined results.
+*/
+
+namespace fc {
+
+namespace impl {
+constexpr static ssize_t append_t = -1;
+
+#ifndef _WIN32
+struct random_access_file_context {
+   random_access_file_context(const random_access_file_context&) = delete;
+   random_access_file_context& operator=(const random_access_file_context&) = delete;
+
+   using native_handle_type = int;
+
+   explicit random_access_file_context(const std::filesystem::path& path) : display_path(path) {
+      int flags = O_RDWR | O_CREAT;
+#if defined(O_CLOEXEC)
+      flags |= O_CLOEXEC;
+#endif
+#if !defined(__linux__)
+      //see pwrite BUGS; we can't set O_APPEND and use pwrite properly on linux. fortunately there is a workaround (see further down)
+      flags |= O_APPEND;
+#endif
+      fd = open(path.c_str(), flags, DEFFILEMODE);
+      FC_ASSERT(fd != -1, "Failed to open ${fn}: ${e}", ("fn", display_path)("e", strerror(errno)));
+#if !defined(O_CLOEXEC)
+      //just swallow errors on this paranoia
+      flags = fcntl(fd, F_GETFD);
+      if(flags != -1)
+         fcntl(fd, F_SETFD, flags | FD_CLOEXEC);
+#endif
+   }
+
+   template<typename MutableBufferSequence>
+   ssize_t read_from(const MutableBufferSequence& mbs, ssize_t offs) {
+      struct iovec iov[IOV_MAX];
+      int i = 0;
+      for(const boost::asio::mutable_buffer& b : mbs) {
+         iov[i].iov_base = b.data();
+         iov[i].iov_len = b.size();
+         if(++i == IOV_MAX)
+            break;
+      }
+      ssize_t red = -1;
+      do {
+         red = preadv(fd, iov, i, offs);
```

Comment:
I admit `red` is a bit of a play on words (that I've done before[^1])

The past tense of "write" would be "wrote" (or maybe "written" also acceptable) and you'll see that here,
https://github.com/AntelopeIO/spring/blob/a29f3f96aa9bfebc3a33eb4c3578bab74bb08527/libraries/libfc/include/fc/io/random_access_file.hpp#L121
I do use `wrote`.

The past tense of "read" is.. "read" (but pronounced differently). I can't name a variable `read` because it's going to conflict with `read()`. So I used `red` here since "red" is pronounced the same as past tense "read".

[^1]: https://github.com/AntelopeIO/spring/blob/2775315b85f3f0e3955f0c74fb0c16cb7e5b422f/libraries/chain/webassembly/runtimes/eos-vm-oc/ipc_helpers.cpp#L48